src/license_guc.c

```diff
@@ -107,279 +101,100 @@ ts_license_enable_module_loading(void)
-		/* filehandle= */ &tsl_handle);
-
-	if (tsl_handle == NULL || tsl_startup_fn == NULL)
-		goto loading_failed;
-
-get_validation_function:
-	tsl_validate_license_fn = lookup_external_function(tsl_handle, "tsl_license_update_check");
-
-	if (tsl_validate_license_fn == NULL)
-		goto loading_failed;
-
-	return true;
-
-/*
- * We want this function to be atomic, either all three relevant values are set
- * or none of them are: If we fail to find one of the values set the two static
- * ones back to NULL. (If they were non-NULL to start with that means the
- * function must have been called successfully in the past, so the lookups
- * should still succeed)
- */
-loading_failed:
-	tsl_handle = NULL;
-	tsl_startup_fn = NULL;
-	tsl_validate_license_fn = NULL;
-	return false;
-}
-
-/* SQL functions */
-
-PGDLLEXPORT Datum
-ts_tsl_loaded(PG_FUNCTION_ARGS)
-{
-	if (TS_CURRENT_LICENSE_IS_APACHE_ONLY())
-		PG_RETURN_BOOL(false);
-
-	PG_RETURN_BOOL(ts_cm_functions->check_tsl_loaded());
-}
-
-PGDLLEXPORT Datum
-ts_enterprise_enabled(PG_FUNCTION_ARGS)
-{
-	if (TS_CURRENT_LICENSE_IS_APACHE_ONLY())
-		PG_RETURN_BOOL(false);
-
-	PG_RETURN_BOOL(ts_cm_functions->enterprise_enabled_internal());
-}
-
-PGDLLEXPORT Datum
-ts_current_license_key(PG_FUNCTION_ARGS)
-{
-	if (!is_member_of_role(GetUserId(), DEFAULT_ROLE_READ_ALL_SETTINGS))
-		ereport(ERROR,
-				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
-				 errmsg("must be superuser or a member of pg_read_all_settings to examine the "
-						"license key")));
-	Assert(ts_guc_license_key != NULL);
-	PG_RETURN_TEXT_P(cstring_to_text(ts_guc_license_key));
-}
-
-PGDLLEXPORT Datum
-ts_print_tsl_license_expiration_info(PG_FUNCTION_ARGS)
-{
-	if (ts_cm_functions->print_tsl_license_expiration_info_hook != NULL)
-		ts_cm_functions->print_tsl_license_expiration_info_hook();
-	PG_RETURN_VOID();
-}
-
-PGDLLEXPORT Datum
-ts_license_expiration_time(PG_FUNCTION_ARGS)
-{
-	if (ts_cm_functions->print_tsl_license_expiration_info_hook == NULL)
-		PG_RETURN_TIMESTAMPTZ(DT_NOEND);
-
-	PG_RETURN_TIMESTAMPTZ(ts_cm_functions->license_end_time());
-}
-
-PGDLLEXPORT Datum
-ts_license_edition(PG_FUNCTION_ARGS)
-{
-	char *edition = NULL;
+	/* If loading is delayed, save the GucSource for later retry
+	 * in the ts_license_enable_module_loading() */
+	if (!load_enabled)
+	{
+		load_source = source;
+		return true;
+	}
 
-	switch (TS_CURRENT_LICENSE_TYPE())
+	if (type == LICENSE_TIMESCALE && !tsl_module_load())
 	{
-		case LICENSE_TYPE_APACHE_ONLY:
-			edition = TS_LICENSE_APACHE_ONLY_PRINT_TEXT;
-			break;
-		case LICENSE_TYPE_COMMUNITY:
-			edition = TS_LICENSE_COMMUNITY_PRINT_TEXT;
-			break;
-		case LICENSE_TYPE_ENTERPRISE:
-			edition = TS_LICENSE_ENTERPRISE_PRINT_TEXT;
-			break;
-		default:
-			elog(ERROR, "Invalid license key '%s'", ts_guc_license_key);
-			pg_unreachable();
+		GUC_check_errdetail("Could not find TSL timescaledb module.");
+		GUC_check_errhint("Check that \"%s\" is available.", EXTENSION_TSL_SO);
+		return false;
```

Comment:
According to Codecov this error condition is not tested.

Reply:
Not sure it is possible to test it now